### ISSUE TITLE
Precision comparison to Python version

### ISSUE THREAD

**Issue by davidmfrey** [PAIR-code/deeplearnjs#765]
I have a model that I trained in Keras w/ tensorflow backend that I am trying to port over to deeplearn.js.  I believe I have the model properly implemented, but it seems there are precision errors that build up as the data goes through the network.  By the final layer, it is rather large, and is throwing off the predictions.  I am comparing the logits at each step with the logits of tensorflow running on cpu.  I'm unsure of the next step in debugging the differences.  I'm hoping someone here might know of any potential precision differences and how I can mitigate them.  I can provide more info as needed.

**Comment by nsthorat** [comment]
Hi @davidmfrey,

Oh no, that's not good. What device are you using? We have precision issues on iOS / Safari right now which we're working on fixing (they don't support floating point textures so we do our own float packing).

Do you have a sense for where things start to diverge, or is it gradual?

Is there a test webpage you could put up that maybe shows the keras results / model, and has deeplearn.js inference and prints of each layer? That would be really helpful!

**Comment by davidmfrey** [comment]
I am using Chrome on iOS.  From what I can tell (assuming Python is printing accurately), there are slight divergences after the first convolutional operation.  It gets gradually worse with each successive layer.  I can try to get the complete data set up somewhere for easy comparison.  The model is 2 LSTM layers which have 10 inputs of a 3 layer convolutional network.

**Comment by nsthorat** [comment]
Can you console log "dl.ENV" after your program has run for a bit and then print the resulting object here?

**Comment by davidmfrey** [comment]
@nsthorat  I'm not sure how much you want to see.  
```
t {features: {…}, BACKEND_REGISTRY: {…}, backends: {…}, currentBackendType: "webgl", globalEngine: t, …}
BACKEND_REGISTRY:{cpu: t, webgl: t}
backends:{cpu: t, webgl: t}
currentBackendType:"webgl"
engine:(...)
features:{DEBUG: false, WEBGL_VERSION: 2, WEBGL_GET_BUFFER_SUB_DATA_ASYNC_EXTENSION_ENABLED: false, BACKEND: "webgl", WEBGL_FLOAT_TEXTURE_ENABLED: true, …}
globalEngine:t {backend: t, customBackend: false, safeMode: false, registeredVariables: {…}, refCounter: WeakMap, …}
globalMath:t {matMul: ƒ, vectorTimesMatrix: ƒ, outerProduct: ƒ, matrixTimesVector: ƒ, dotProduct: ƒ, …}
math:(...)
__proto__:Object
```

I also have the output of a JSON.stringify on it, but it's 2 MB.  I can email or share a link if it would help.

Also forgot to mention that I'm using 
  <script src="https://cdn.jsdelivr.net/npm/deeplearn@latest"></script>
to load the script, so I'm up to date.

Thanks for your help

**Comment by davidmfrey** [comment]
Just another quick note, I tried with calling dl.setBackend('cpu') first and it's generating the same incorrect output. I'll try to store the output of some layers tomorrow for comparison.

**Comment by nsthorat** [comment]
How are you doing the conversion from keras => deeplearn? Are you sure your paddings / all the arguments to conv2d line up? The shapes also line up? And making sure you're using the right conv2d (not depthwise, etc)?

**Comment by davidmfrey** [comment]
Good question... I've had to write them manually, especially the LSTM as Keras has activations applied.   Here's the deeplearn model.  I haven't moved it to the graph/session api yet. The weight and bias vectors all look identical in JS and Python.
```
class Cnn {
  constructor( weights, biases ) {
    this._weights = weights;
    this._biases = biases;
  }

  call(input) {
    return dl.tidy(() => {
      let tensor = dl.tensor4d(input, [1,240,240,1]);
      tensor = tensor.conv2d(this._weights[0], 1, 'same');
      tensor = tensor.add(this._biases[0]);
      tensor = tensor.relu();
      tensor = tensor.maxPool(4, 4, 'same');
      tensor = tensor.conv2d(this._weights[1], 1, 'same');
      tensor = tensor.add(this._biases[1]);
      tensor = tensor.relu();
      tensor = tensor.maxPool(3, 3, 'same');
      tensor = tensor.conv2d(this._weights[2], 1, 'same');
      tensor = tensor.add(this._biases[2]);
      tensor = tensor.relu();
      tensor = tensor.maxPool(4, 4, 'same');
      tensor = tensor.flatten();
      return tensor;
    });
  }
}
```
And the Keras definition
```
cnn_model = Sequential([
    Conv2D(filters=8, kernel_size=3, padding='same', activation='relu', input_shape=(240,240,1)),
    MaxPooling2D(pool_size=4),
    Conv2D(filters=16, kernel_size=3, padding='same', activation='relu'),
    MaxPooling2D(pool_size=3),
    Conv2D(filters=32, kernel_size=3, padding='same', activation='relu'),
    MaxPooling2D(pool_size=4),
    Flatten()])
```
I can post the LSTM as well, but the outputs of the CNNs are slightly different already.  It just gets more exaggerated as it goes through the LSTM.

Here's an example of the values coming out of the CNN

JS
```
[0, 0, 0, 0, 0, 0, 0, 0.05034855008125305, 0, 0, 0, 0, 0, 0, 0, 0.14068517088890076, 0, 0, 0.1277533322572708, 0, 0, 0.32633066177368164, 0, 0, 0.40381020307540894, 0.020297199487686157...
```
Python
```
[0, 0, 0, 0, 0, 0, 0, 0.0503485203, 0, 0, 0, 0, 0, 0, 0, 0.140685171, 0, 0, 0.127753362, 0, 0, 0.3263309, 0, 0, 0.403810203, 0.0202972591...
```

As I said, the weights look exactly the same (rounded up/down in Python when printing fewer decimal places)

Python example weights
```
[0.0452684946 -0.203179061 -0.120704241 -0.115379043 -0.0178494938...
```
JS weights
```
[0.04526849463582039, -0.20317906141281128, -0.12070424109697342, -0.11537904292345047, -0.017849493771791458...
```

**Comment by davidmfrey** [comment]
@nsthorat  Apologies, I'm traveling this weekend.  It will probably be Monday or Tuesday before I can get any farther on this.

**Comment by davidmfrey** [comment]
@nsthorat, I retrained my model with a simple rnn implementation and it is now working correctly.  The problem must be in the LSTM implementation I copied from Keras.  Sorry for the confusion and thanks for your patience.

**Comment by nsthorat** [comment]
No problem at all :)
